### PR TITLE
Plugins: Always load decoupled frontend assets from builds

### PR DESCRIPTION
**Pull request by jackw**
**What is this feature?**

This PR separates frontend builds of core decoupled datasource plugins from Grafana core. If a plugin is "decoupled" it will always be loaded via the `dist/module.js` file that is created by building the datasource separately.

**Why do we need this feature?**

Right now decoupled plugins continue to be bundled in grafana core codebase. For "decoupled" plugin developers this means they _should_ be regularly checking both bundled and decoupled frontend builds. The changes in this PR give us a single way to load decoupled plugins which should alleviate this. 

To help development we're using NX to orchestrate building all external plugins when running `yarn start`. These build tasks are cached which should speed things up, particularly for core Grafana contributors, as the separation means less code for webpack to churn through. 

image

The above screenshot highlights that with the changes in this PR reduce dev start time for a cached webpack build from 15secs to 6.5secs. As more decoupling happens this number will continue to shrink. 🚀 

However this changes workflows.. decoupled plugin devs now need to (potentially) run two commands to develop. One for building grafana core and one for building their plugin.

- To build all plugins and watch grafana core for changes -> `yarn start`
- To build and watch all plugins -> `yarn plugin:build:dev`
- To build and watch a single plugin -> `yarn workspace @grafana-plugins/grafana-testdata-datasource dev`

Whilst this might seem inconvenient they run in parallel and a plugin developer doesn't need to wait for all externalised plugins to build when making changes to their own plugin. If the work is only focused on the plugin you might not even need to run `yarn start`.

I believe doing this will streamline decoupled datasource development in the frontend and move us further towards a frontend mono-repo where we focus on building only the necessary pieces rather than the entire jungle. We also start to see some local benefits from using NXs cache feature.

- [x] Update backend to always load decoupled datasources from dist/
- [x] Update frontend to not bundle decoupled datasources
- [x] Implement nx for task orchestration and explicit dependencies
- [x] Ignore watching decoupled plugins in core webpack config

**Who is this feature for?**

Grafana developers

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**
This does NOT enable any caching in CI. We should aim to do that iteratively, making sure params are set to invalidate caches so we don't end up with cases where builds should fail but the cache was used and they passed.

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: package.json
Comment (joshhunt) [inline package.json:75]: this can't live in the nx.json file? Weird that we have/need both.

File: package.json
Comment (itsmylife) [inline package.json:64]: We should add the Prometheus package here when it has a build step, shouldn't we?

File: package.json
Comment (jackw) [reply package.json:75]: I had some weirdness with running top level scripts with NX and this seemed like the more reliable place for it to work. I'll try fiddling some more as I'd also like it all in a single place. 👍

File: package.json
Comment (jackw) [reply package.json:64]: Exactly! I'd like us to reach consensus first on this approach before updating our google doc to highlight what needs to be done where to make sure plugins are decoupled without issue. 👍

**Comment by andresmgot** [comment]
I tested this and it's working as expected so my +1 for the approach. In any case, feels like we should put this behind a feature flag so we can properly test it before moving it to production. Since this is touching how the plugins are loaded, it can be tricky though.

> * [ ]  Introduce new CI step to build decoupled frontend plugins

Why shouldn't we keep the previous behavior of building the plugins as well when running the `build` task? It's convenient. If we change it and require two different commands, we also need to update the developer documentation, basically everybody will need to build the plugins that they plan to use.

**Review by itsmylife — APPROVED** [review]

**Comment by jackw** [comment]
>Why shouldn't we keep the previous behavior of building the plugins as well when running the build task? It's convenient. If we change it and require two different commands, we also need to update the developer documentation, basically everybody will need to build the plugins that they plan to use.

I'm kind of on the fence about the idea too and happy to leave it as it / update the build task to build all the plugins. My reasoning was it'll be faster in CI and easier to identify failures if we build these pieces (core frontend and plugins) separately. If we go for this option with CI for local development we could have script "build:all" for developers that want to do prod builds locally. Or flip that on it's head and have "build:core" that only builds grafana frontend and leave "build" as is. WDYT?

**Comment by andresmgot** [comment]
> Or flip that on it's head and have "build:core" that only builds grafana frontend and leave "build" as is. WDYT?

I like that more to avoid the "breaking change"

**Review by wbrowne — APPROVED** [review]
Backend changes look good to me! Awesome work 🚀

**Comment by jackw** [comment]
> > Or flip that on it's head and have "build:core" that only builds grafana frontend and leave "build" as is. WDYT?
> 
> I like that more to avoid the "breaking change"

@andresmgot after discussing with release-guild members it seems it will be rather problematic to change the frontend build command in CI so I'll be working towards your original suggestion to not create breaking changes there. 👍

File: package.json
Comment (academo) [inline package.json:61]: I vaguely remember the NX config by now, but can we move this to the NX.json file instead?

**Review by academo — COMMENTED** [review]
This looks great @jackw but we are missing the documentation aspect of this setup. 

Can you add a section to the readme (link to a new file) about how to proceed when decoupling a plugin from core? We'll keep working in more decoupling in the future and I am afraid people won't know all the files they have to modify to make sure their plugin is in the right setup

**Comment by academo** [comment]
> There is a live [doc](https://docs.google.com/document/d/1Fh6NeI7mNQwmtzPMhawMnmHypWd_PD-ZJQ-bDseRy1g/edit#heading=h.4cli68q9dq8p) for this.

Can we link it here? It seems hard to discover (since I could not find it)

File: package.json
Comment (jackw) [reply package.json]: Because of the way top-level scripts work with NX it doesn't seem we can put this info in the nx.json file. I've moved it over to a project.json file and cleaned up the repetition.

File: package.json
Comment (jackw) [reply package.json]: Because of the way top-level scripts work with NX it doesn't seem we can put this info in the nx.json file. I've moved it over to a project.json file and cleaned up the repetition.

File: package.json
Comment (jackw) [reply package.json]: Reading this again, I thought you were refering to the extracted prometheus plugin not the package. The package won't need to be added here as this repository consumes its src files. Build is only required for publishing a package to the npm registry.

**Comment by jackw** [comment]
@academo thanks for the feedback. I've posted this in a couple of places in slack asking for feedback as this is still a proposal. Once we've got approval on this direction we'll be sure to update the doc that explains the necessary steps required to decouple a plugin. 👍

File: scripts/webpack/webpack.dev.js
Comment (andresmgot) [inline scripts/webpack/webpack.dev.js:43]: this list seems incomplete (compared to https://github.com/grafana/grafana/blob/main/.eslintrc#L98-L119). Maybe we can read / adapt that list here?

Comment (jackw) [reply scripts/webpack/webpack.dev.js:43]: I'll need to dig into what's going on in that eslintrc. I only consider a decoupled datasource when it has a package.json and according to my IDE if I'm on `main` I only see the following. 🤔 

image

Comment (andresmgot) [reply scripts/webpack/webpack.dev.js:43]: yeah, plugins can have the linter rules active but still not be fully decoupled. Then, can we add some logic here to find package.jsons? I am trying to avoid a manual step for people that will likely forget.

**Review by andresmgot — APPROVED** [review]
Did some basic testing and everything seems good!

File: scripts/webpack/webpack.dev.js
Comment (jackw) [reply scripts/webpack/webpack.dev.js]: Nice suggestion @andresmgot ! I've added something in https://github.com/grafana/grafana/pull/81873/commits/7584e613631abc86941708ee91ebe246fde5a224 which should address this without the need for additional steps during decoupling.

Comment (andresmgot) [reply scripts/webpack/webpack.dev.js]: :clap:

**Review by bohandley — APPROVED** [review]
LGTM! No concerns here, thank you for this work!

**Comment by torkelo** [comment]
@jackw this looks great! definitely the feels like the right direction (speeding up builds for individual parts/plugins) , and making this more uniform

**Review by torkelo — APPROVED** [review]

**Review by andresmgot — COMMENTED** [review]
Just a note that Zipkin is now decoupled too :)

**Review by Clarity-89 — APPROVED** [review]
Great work on this 🚀  Just to double check - this (mainly the usage of NX) doesn't impact the way grafana packages (e.g. grafana/ui) are versioned and published?

File: package.json
Comment (jackw) [reply package.json:42]: It's also a convention in Nx which uses the colon to delimit `nx run {project}:{target}:{configuration}`, so it can interpret the command as a target and configuration. When this happens it can fail to run. A workaround is to wrap the command in quotes but that (IMHO) gets messy. I'm not sure this change is necessary for what I'm proposing in this PR but running the command individually like `yarn nx run grafana:themes:generate` might fail.

File: package.json
Comment (Clarity-89) [reply package.json:60]: I was thinking about having smth like `nx.config.base.json`, which can have the basic config options:
```json
{
  "projects": "@grafana-plugins/*"
  "exclude": "@grafana-plugins/input-datasource"
}
```
and then use it `"plugin:build": "nx run-many -t build --config=nx.config.base.json`. Other build configs would extend it, kinda similar to how webpack configs are structured. But again, not sure if anything like this is possible with Nx.

**Comment by jackw** [comment]
> Great work on this 🚀  Just to double check - this (mainly the usage of NX) doesn't impact the way grafana packages (e.g. grafana/ui) are versioned and published? 

No Nx is a dependency of Lerna solely for task orchestration. We will continue to use Lerna for versioning and publishing as we always have.

**Pull request merged**
